src/transport/virtiofs/mod.rs

```diff
@@ -105,10 +105,11 @@ impl std::error::Error for Error {}
 
 impl<'a> Reader<'a> {
     /// Construct a new Reader wrapper over `desc_chain`.
-    pub fn new(
-        mem: &'a GuestMemoryMmap,
-        desc_chain: DescriptorChain<GuestMemoryAtomic<GuestMemoryMmap>>,
-    ) -> Result<Reader<'a>> {
+    pub fn new<M>(mem: &'a GuestMemoryMmap, desc_chain: DescriptorChain<M>) -> Result<Reader<'a>>
```

Comment:
Should we use "mem: M" instead of "mem: &'a GuestMemoryMmap"?

Reply:
Do you mean adding `M: Deref + GuestMemory`? vhost-user-backend requires `GuestMemoryAtomic<GuestMemoryMmap>` but I don't see how to convert it back to `GuestMemory`. Looks like vm-memory is missing some abstractions, e.g.:

```
the trait `vm_memory::GuestMemory` is not implemented for `Arc<GuestMemoryMmap>`
```

Reply:
Does this work?
```
pub fn new<M>(mem: M, desc_chain: DescriptorChain<M>) -> Result<Reader<'a>> 
    where
        M: Deref,
        M::Target: GuestMemory + Sized,
    {
```

Reply:
In that case, I'm getting:

> 
   --> src/transport/virtiofs/mod.rs:136:14
    |
136 |             .collect::<Result<VecDeque<VolatileSlice<'a>>>>()?;
    |              ^^^^^^^ value of type `VecDeque<VolatileSlice<'a>>` cannot be built from `std::iter::Iterator<Item=VolatileSlice<'_, <<<<M as std::ops::Deref>::Target as vm_memory::GuestMemory>::R as GuestMemoryRegion>::B as WithBitmapSlice<'_>>::S>>`

Reply:
then keep it as is, it will be solve by another way.